### DISCUSSION TITLE
Gitea abstraction and altenatives

### DISCUSSION THREAD

**Discussion by liamfallon** [nephio-project/nephio#424] [Development]
I was looking at writing some unit test for the repository reconciler and was trying to mock the Gitea client. It seems to use the Gitea client directly. See:
https://github.com/nephio-project/nephio/tree/main/controllers/pkg/reconcilers/repository

This makes it pretty awkward to mock, as to mock it we would need to abstract the Gitea client structure into an interface. This then would have overlaps with work to allow other repo systems besides Gitea.

Other people are looking at Gitea and I see the Gitea client has an interface in it. I wonder if anyone had any thoughts on this?

**Comment by vjayaramrh** [comment]
@liamfallon What tool are you using for creating mocks?

**Reply by liamfallon** [reply]
I couldn't see anything really suitable. I was going to do the mocks manually for now.

**Reply by liamfallon** [reply]
Maybe the community has some recommendations?

**Reply by vjayaramrh** [reply]
I have used mockery in the past https://vektra.github.io/mockery/latest/#who-uses-mockery. We can jointly evaluate this framework, let me know.
cc: @efiacor

**Reply by liamfallon** [reply]
I'll have a look at this, thanks @vjayaramrh

**Comment by henderiw** [comment]
one thing I always wanted to do is this:
- we have the respective git wrappers: like GitHub, gitea, etc
- we have the way we interact with git itself. For the latter there is https://github.com/go-git/go-git

For the git wrappers I have not found anything but we could build something that helps us interact with the various git providers.

**Reply by liamfallon** [reply]
That sounds like a good approach. I guess we would have a common interface that the resoolvers would use and then a different implementation of that interface for gitea, git or whatever and also a dummy implementation for mocking in unit tests.

**Comment by liamfallon** [comment]
I looked a the Mockery framework suggested by @vjayaramrh. I think we should use it where possible for our unit tests.
